src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/TimeOnlyConverter.cs

```diff
@@ -3,30 +3,72 @@
 
 using System.Buffers.Text;
 using System.Diagnostics;
-using System.Globalization;
 
 namespace System.Text.Json.Serialization.Converters
 {
     internal sealed class TimeOnlyConverter : JsonConverter<TimeOnly>
     {
-        private static readonly TimeSpanConverter s_timeSpanConverter = new TimeSpanConverter();
-        private static readonly TimeSpan s_timeOnlyMaxValue = TimeOnly.MaxValue.ToTimeSpan();
+        private const int MinimumTimeOnlyFormatLength = 8; // hh:mm:ss
+        private const int MaximumTimeOnlyFormatLength = 16; // hh:mm:ss.fffffff
+        private const int MaximumEscapedTimeOnlyFormatLength = JsonConstants.MaxExpansionFactorWhileEscaping * MaximumTimeOnlyFormatLength;
 
         public override TimeOnly Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
         {
-            TimeSpan timespan = s_timeSpanConverter.Read(ref reader, typeToConvert, options);
+            if (reader.TokenType != JsonTokenType.String)
+            {
+                ThrowHelper.ThrowInvalidOperationException_ExpectedString(reader.TokenType);
+            }
 
-            if (timespan < TimeSpan.Zero || timespan > s_timeOnlyMaxValue)
+            if (!JsonHelpers.IsInRangeInclusive(reader.ValueLength, MinimumTimeOnlyFormatLength, MaximumEscapedTimeOnlyFormatLength))
             {
-                ThrowHelper.ThrowJsonException();
+                ThrowHelper.ThrowFormatException(DataType.TimeOnly);
             }
 
+            ReadOnlySpan<byte> source = stackalloc byte[0];
+            if (!reader.HasValueSequence && !reader.ValueIsEscaped)
+            {
+                source = reader.ValueSpan;
+            }
+            else
+            {
+                Span<byte> stackSpan = stackalloc byte[MaximumEscapedTimeOnlyFormatLength];
```

Comment:
CopyString copies unescaped string (at least per comment https://github.com/eiriktsarpalis/runtime/blob/9c6ded4b654a1abb18aab44e33d469d29393ea5f/src/libraries/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs#L51), why do we need to allocate for max escaped rather than unescaped?

Reply:
It's because we don't know what the actual unescaped size is ahead of time. It might be that the source buffer has length equal to `MaximumEscapedTimeOnlyFormatLength` but none of the characters are escaped (and therefore must be invalid). In this case we'd rather the input be rejected by the parser than have the `CopyString` method fail with a `destinationTooSmall` exception.